### PR TITLE
fix(compiler): write exports for defineCustomElement typedefs

### LINKED ISSUE #4093: bug: `single-export-module` is missing types for `defineCustomElement` functions
### Prerequisites

- [X] I have read the [Contributing Guidelines](https://github.com/ionic-team/stencil/blob/main/.github/CONTRIBUTING.md).
- [X] I agree to follow the [Code of Conduct](https://github.com/ionic-team/stencil/blob/main/CODE_OF_CONDUCT.md).
- [X] I have searched for [existing issues](https://github.com/ionic-team/stencil/issues) that already report this problem, without success.

### Stencil Version

3.0.1

### Current Behavior

When using the `dist-custom-element` output target with `customElementsExportBehavior: 'single-export-module'` we are experiencing missing type declarations for the `defineCustomElement...` functions of our components.

```
export { MyComponent as MyComponent } from '../types/components/my-component/my-component';
```

### Expected Behavior

When publishing as a component library, I'd expect that consumers can access the type for the `defineCustomElement...` function the same way they can access the component's type.

```
export { MyComponent as MyComponent, defineCustomElement as defineCustomElementMyComponent  } from '../types/components/my-component/my-component';
```

### System Info

_No response_

### Steps to Reproduce

I've created a stackblitz reproduction example here: https://stackblitz.com/edit/node-mhjzxw?file=stencil.config.ts,dist%2Fcomponents%2Findex.js,dist%2Fcomponents%2Findex.d.ts

### Code Reproduction URL

https://stackblitz.com/edit/node-mhjzxw?file=stencil.config.ts,dist%2Fcomponents%2Findex.js,dist%2Fcomponents%2Findex.d.ts

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by tanner-reits**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [X] Build (`npm run build`) was run locally and any changes were pushed
- [X] Unit tests (`npm test`) were run locally and passed
- [X] E2E Tests (`npm run test.karma.prod`) were run locally and passed
- [X] Prettier (`npm run prettier`) was run locally and passed

## Pull request type

 

 

Please check the type of change your PR introduces:
- [X] Bugfix
- [ ] Feature
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

The `index.d.ts` file for `dist-custom-elements` when setting `generateTypeDeclarations: true` and `customElementsExportBehavior: 'single-export-module'` would not include exported typedefs for the components' `defineCustomElement` functions.

GitHub Issue Number: Closes #4093 

## What is the new behavior?

The `index.d.ts` file now includes an export for `defineCustomElement` for each component in the bundle

## Does this introduce a breaking change?

- [ ] Yes
- [X] No

## Testing

Verified output in a Stencil starter project and in a consuming application (Angular)

Updated unit tests for the affected code.

### CODE REVIEW

**Comment by alicewriteswrongs** [comment]
looks there might be one additional SNC error though - would you mind fixing that while we're in here?

File: src/compiler/output-targets/dist-custom-elements/custom-elements-types.ts [hunk 1]
```diff
@@ -75,7 +75,16 @@ const generateCustomElementsTypesOutput = async (
             const componentSourceRelPath = relative(config.srcDir, component.sourceFilePath).replace('.tsx', '');
             const componentDTSPath = join(componentsTypeDirectoryRelPath, componentSourceRelPath);
 
-            return `export { ${importName} as ${exportName} } from '${componentDTSPath}';`;
+            const defineFunctionExportName = `defineCustomElement${exportName}`;
+            // Get the path to the sibling typedef file for the current component
+            const localComponentTypeDefFilePath = `./${component.sourceFilePath.split('/').pop()?.replace('.tsx', '')}`;
```
Comment (rwaskiewicz) [inline src/compiler/output-targets/dist-custom-elements/custom-elements-types.ts:80]: Can we tighten this replace up by ensuring we only replace '.tsx' if it occurs at the end of the string? I think that'd look something like:
```suggestion
            const localComponentTypeDefFilePath = `./${component.sourceFilePath.split('/').pop()?.replace(/\.tsx$/, '')}`;
```

**Review by rwaskiewicz — CHANGES_REQUESTED** [review]

**Comment by tanner-reits** [comment]
Made a small change to how the path is generated to the sibling typedef file since when we bundle the code it generates the `.js` and `.d.ts` files from the component tag name _not_ the file name

I did, however, update how we replace the `.tsx` extension for the path to the `dist-types` file so it only strips off the trailing `.tsx` extension (similar to [this comment](https://github.com/ionic-team/stencil/pull/4194#discussion_r1150593965))

**Review by alicewriteswrongs — APPROVED** [review]
lgtm! 👍

**Review by rwaskiewicz — APPROVED** [review]

**Pull request merged**